### PR TITLE
 Profiling tool: Update comparison mode output format and add error handling

### PR DESCRIPTION
**Pull request by nartal1**
This PR fixes multiple FEA/bugs. 
1. Added appName to profiling output, fixes https://github.com/NVIDIA/spark-rapids/issues/2742
2. Added error handling for comparison functions, fixes https://github.com/NVIDIA/spark-rapids/issues/2692
3. Updated output for comparison mode functions, fixes https://github.com/NVIDIA/spark-rapids/issues/2685
Updated the tests and README to relfect these changes.

After adding appName, output will be as below:
```
+--------+-----------+------------------------------+-------------+-------------+--------+-----------+------------+-------+
|appIndex|appName    |appId                         |startTime    |endTime      |duration|durationStr|sparkVersion|gpuMode|
+--------+-----------+------------------------------+-------------+-------------+--------+-----------+------------+-------+
|1       |Spark shell|application_1616728744956_0028|1616988233052|1616988390501|157449  |2.6 min    |3.0.1       |true   |
|2       |Spark shell|application_1616728744956_0009|1616740267325|1616740893089|625764  |10 min     |3.0.1       |true   |
|3       |Spark shell|application_1616746343401_0025|1617764543255|1617764863521|320266  |5.3 min    |3.0.1       |true   |

```

   
For comparison mode, output is updated for below functions:
    a. Executor Information:
```
+--------+------------+----------------+-----------+------------+-------------+--------+--------+--------+------------+--------+--------+
|appIndex|numExecutors|coresPerExecutor|maxMemory  |maxOnHeapMem|maxOffHeapMem|exec_cpu|exec_mem|exec_gpu|exec_offheap|task_cpu|task_gpu|
+--------+------------+----------------+-----------+------------+-------------+--------+--------+--------+------------+--------+--------+
|1       |2           |8               |47055896576|25581060096 |21474836480  |null    |null    |null    |null        |null    |null    |
|2       |2           |8               |55645831168|12696158208 |42949672960  |null    |null    |null    |null        |null    |null    |
|3       |24          |4               |42760929280|25581060096 |17179869184  |null    |null    |null    |null        |null    |null    |
|4       |24          |16              |42760929280|25581060096 |17179869184  |null    |null    |null    |null        |null    |null    |
+--------+------------+----------------+-----------+------------+-------------+--------+--------+--------+------------+--------+--------+
```

   b. Rapids properties which are explicitly set
```
+------------------------------------------------------------+----------+----------+-------------+-------------+
|propertyName                                                |appIndex_1|appIndex_2|appIndex_3   |appIndex_4   |
+------------------------------------------------------------+----------+----------+-------------+-------------+
|spark.rapids.memory.gpu.debug                               |STDERR    |STDERR    |STDERR       |STDERR       |
|spark.rapids.memory.gpu.pooling.enabled                     |true      |true      |true         |true         |
|spark.rapids.memory.pinnedPool.size                         |16G       |16G       |16G          |16G          |
|spark.rapids.sql.batchSizeBytes                             |1073741824|1073741824|50331648     |null         |
|spark.rapids.sql.castFloatToString.enabled                  |true      |true      |true         |true         |

```

   c. SQL Duration and Executor CPU Time Percent - (sorted order based on appIndex)
```
+--------+------------------------------+-----+------------+-------------------+------------+------------------+-------------------------+
|appIndex|App ID                        |sqlID|SQL Duration|Contains Dataset Op|App Duration|Potential Problems|Executor CPU Time Percent|
+--------+------------------------------+-----+------------+-------------------+------------+------------------+-------------------------+
|1       |application_1616728744956_0028|1    |18926       |false              |157449      |null              |34.15                    |
|1       |application_1616728744956_0028|0    |57854       |false              |157449      |null              |37.3                     |
|1       |application_1616728744956_0028|2    |14034       |false              |157449      |null              |38.53                    |
|2       |application_1616728744956_0009|0    |62433       |false              |625764      |null              |43.04                    |
|2       |application_1616728744956_0009|2    |15844       |false              |625764      |null              |23.32                    |
|2       |application_1616728744956_0009|1    |24855       |false              |625764      |null              |20.02                    |
|3       |application_1616746343401_0025|0    |259949      |false              |320266      |null              |54.21                    |
|4       |application_1616746343401_0028|2    |8           |false              |1559906     |null              |null                     |
|4       |application_1616746343401_0028|1    |1           |false              |1559906     |null              |null                     |

```

  d. Failed tasks, jobs, stages include appIndex for better mapping:

```
+--------+-------+--------------+------+-------+----------------------------------------------------------------------------------------------------+
|appIndex|stageId|stageAttemptId|taskId|attempt|failureReason                                                                                       |
+--------+-------+--------------+------+-------+----------------------------------------------------------------------------------------------------+
|3       |4      |0             |2842  |0      |ExceptionFailure(ai.rapids.cudf.CudfException,cuDF failure at: /home/jenkins/agent/workspace/jenkins|
|3       |4      |0             |2858  |0      |TaskKilled(another attempt succeeded,List(AccumulableInfo(453,None,Some(22000),None,false,true,None)|
|3       |4      |0             |2884  |0      |TaskKilled(another attempt succeeded,List(AccumulableInfo(453,None,Some(21148),None,false,true,None)|
|3       |4      |0             |2908  |0      |TaskKilled(another attempt succeeded,List(AccumulableInfo(453,None,Some(20420),None,false,true,None)|

```

### CODE REVIEW

File: tools/README.md [hunk 1]
```diff
@@ -241,29 +241,29 @@ We can input multiple Spark event logs and this tool can compare environments, e
...
 - Compare Executor information:
 ```
 Compare Executor Information:
-+--------+----------+----------+-----------+------------+-------------+--------+--------+--------+------------+--------+--------+
-|appIndex|executorID|totalCores|maxMem     |maxOnHeapMem|maxOffHeapMem|exec_cpu|exec_mem|exec_gpu|exec_offheap|task_cpu|task_gpu|
-+--------+----------+----------+-----------+------------+-------------+--------+--------+--------+------------+--------+--------+
-|1       |0         |4         |13984648396|13984648396 |0            |null    |null    |null    |null        |null    |null    |
-|1       |1         |4         |13984648396|13984648396 |0            |null    |null    |null    |null        |null    |null    |
++--------+------------+----------------+-----------+------------+-------------+--------+--------+--------+------------+--------+--------+
+|appIndex|numExecutors|coresPerExecutor|maxMem     |maxOnHeapMem|maxOffHeapMem|exec_cpu|exec_mem|exec_gpu|exec_offheap|task_cpu|task_gpu|
```
Comment (tgravescs) [inline tools/README.md:256]: we need to print resource profile id, I think we should also clarify the other columns.. exec_gpu for instance perhaps that is just numGpusPerExecutor or something more readable.  how is exec_cpu different from totalCores?  Perhaps we need to make it more clear about resource profiles only being parsable when processing with spark 3.1+ as well.

File: tools/README.md [hunk 2]
```diff
@@ -241,29 +241,29 @@ We can input multiple Spark event logs and this tool can compare environments, e
...
 Compare Application Information:
 
-+--------+-----------------------+-------------+-------------+--------+-----------+------------+-------+
-|appIndex|appId                  |startTime    |endTime      |duration|durationStr|sparkVersion|gpuMode|
-+--------+-----------------------+-------------+-------------+--------+-----------+------------+-------+
-|1       |app-20210329165943-0103|1617037182848|1617037490515|307667  |5.1 min    |3.0.1       |false  |
-|2       |app-20210329170243-0018|1617037362324|1617038578035|1215711 |20 min     |3.0.1       |true   |
-+--------+-----------------------+-------------+-------------+--------+-----------+------------+-------+
++--------+-----------+-----------------------+-------------+-------------+--------+-----------+------------+-------+
+|appIndex|appName    |appId                  |startTime    |endTime      |duration|durationStr|sparkVersion|gpuMode|
```
Comment (tgravescs) [inline tools/README.md:245]: perhaps we should change gpuMode to Plugin Enabled or something -> note we can do this as separate issue just noticed here.

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/CompareApplications.scala [hunk 3]
```diff
@@ -36,15 +36,19 @@ class CompareApplications(apps: ArrayBuffer[ApplicationInfo],
     var query = ""
     var i = 1
     for (app <- apps) {
-      query += app.generateAppInfo
-      if (i < apps.size) {
-        query += "\n union \n"
+      if (app.allDataFrames.contains(s"appDF_${app.index}")) {
+        query += app.generateAppInfo
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/CompareApplications.scala:40]: so this really is the same as printAppInfo which the exception of the union vs printing new tables on each iteration.  I don't think in non-compare mode we ever have more then 1 app,b ut seems like we can combine and just have parameter to change behavior.  That might be the same for other functions in this file like job info, etc. so perhaps we can look through all of them and combine where possible

File: tools/src/main/scala/org/apache/spark/sql/rapids/tool/profiling/ApplicationInfo.scala [hunk 4]
```diff
@@ -579,49 +579,75 @@ class ApplicationInfo(
     if (allDataFrames.contains(s"blockManagersDF_$index") &&
         allDataFrames.contains(s"resourceProfilesDF_$index")) {
 
-      s"""select $index as appIndex, e.executorID, e.totalCores,
-         |b.maxMem, b.maxOnHeapMem,b.maxOffHeapMem,
-         |r.exec_cpu, r.exec_mem, r.exec_gpu, r.exec_offheap, r.task_cpu, r.task_gpu
+      s"""select $index as appIndex,
+         |first(r.id) as resourceProfileId,
+         |count(e.executorID) as numExecutors,
```
Comment (tgravescs) [inline tools/src/main/scala/org/apache/spark/sql/rapids/tool/profiling/ApplicationInfo.scala:584]: this isn't unique per resource profile id, which we want

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/ProfileMain.scala [hunk 5]
```diff
@@ -141,7 +141,7 @@ object ProfileMain extends Logging {
       if (appArgs.compare()) { // Compare Applications
 
         textFileWriter.write("### A. Compare Information Collected ###")
-        val compare = new CompareApplications(apps, textFileWriter)
+        val compare = new CompareApplications(apps, Some(textFileWriter))
```
Comment (tgravescs) [inline tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/ProfileMain.scala:144]: so are we printing all the health checks in different tables when doing comapre mode?  I thought we were going to put those into a single table which multiple app index, it seem inconsistent right now so I think we should make ti consistent

File: tools/README.md [hunk 6]
```diff
@@ -370,7 +370,7 @@ SQL Duration and Executor CPU Time Percent
 ```
 Shuffle Skew Check: (When task's Shuffle Read Size > 3 * Avg Stage-level size)
 +--------+-------+--------------+------+-------+---------------+--------------+-----------------+----------------+----------------+----------+----------------------------------------------------------------------------------------------------+
-|appIndex|stageId|stageAttemptId|taskId|attempt|taskDurationSec|avgDurationSec|taskShuffleReadMB|avgShuffleReadMB|taskPeakMemoryMB|successful|endReason_first100char                                                                              |
+|appIndex|stageId|stageAttemptId|taskId|attempt|taskDurationSec|avgDurationSec|taskShuffleReadMB|avgShuffleReadMB|taskPeakMemoryMB|successful|reason                                                                              |
```
[the remark this answers is not in the document]
Comment (nartal1) [reply tools/README.md:373]: yes, it is still 100 char limit. Will update in the readme.

File: tools/src/main/scala/org/apache/spark/sql/rapids/tool/profiling/ApplicationInfo.scala [hunk 7]
```diff
@@ -579,49 +579,75 @@ class ApplicationInfo(
     if (allDataFrames.contains(s"blockManagersDF_$index") &&
         allDataFrames.contains(s"resourceProfilesDF_$index")) {
 
-      s"""select $index as appIndex, e.executorID, e.totalCores,
-         |b.maxMem, b.maxOnHeapMem,b.maxOffHeapMem,
-         |r.exec_cpu, r.exec_mem, r.exec_gpu, r.exec_offheap, r.task_cpu, r.task_gpu
+      s"""select $index as appIndex,
+         |first(r.id) as resourceProfileId,
```
[the remark this answers is not in the document]
Comment (nartal1) [reply tools/src/main/scala/org/apache/spark/sql/rapids/tool/profiling/ApplicationInfo.scala:583]: The eventlog which i had to print the final result was from Spark-3.0.1, so resource profile id is missing. Yes, i see there could be some confusion with the output when the eventlogs are from before Spark-3.1

File: tools/README.md (hunk 1 above)
Comment (nartal1) [reply tools/README.md]: Updated.  We are printing resource profile id only if we are processing Spark3.1+ event logs.

File: tools/src/main/scala/org/apache/spark/sql/rapids/tool/profiling/ApplicationInfo.scala (hunk 7 above)
[the remark this answers is not in the document]
Comment (nartal1) [reply tools/src/main/scala/org/apache/spark/sql/rapids/tool/profiling/ApplicationInfo.scala]: added test for multiple resource profile ids.

File: tools/README.md (hunk 2 above)
Comment (nartal1) [reply tools/README.md]: done

File: tools/src/main/scala/org/apache/spark/sql/rapids/tool/profiling/ApplicationInfo.scala (hunk 4 above)
Comment (nartal1) [reply tools/src/main/scala/org/apache/spark/sql/rapids/tool/profiling/ApplicationInfo.scala]: Good catch. I was not testing this on event logs with mutliple resource ids. Updated the query which may not be efficient but works as intended. My SQL skills isn't that good. Please take a look at this.

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/ProfileMain.scala (hunk 5 above)
Comment (nartal1) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/ProfileMain.scala:123]: Updated. It prints in a single table with multiple app index.

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/Analysis.scala [hunk 8]
```diff
@@ -172,7 +172,7 @@ class Analysis(apps: ArrayBuffer[ApplicationInfo], fileWriter: Option[ToolTextFi
       }.map( app => "(" + app.profilingDurationSQL + ")")
       .mkString(" union ")
     if (query.nonEmpty) {
-      apps.head.runQuery(query, false, fileWriter, messageHeader)
+      apps.head.runQuery(query + "order by appIndex", false, fileWriter, messageHeader)
```
[the remark this answers is not in the document]
Comment (nartal1) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/Analysis.scala]: done

**Comment by nartal1** [comment]
build

File: tools/README.md [hunk 9]
```diff
@@ -245,29 +245,30 @@ We can input multiple Spark event logs and this tool can compare environments, e
...
 - Compare Executor information:
 ```
 Compare Executor Information:
-+--------+----------+----------+-----------+------------+-------------+--------+--------+--------+------------+--------+--------+
-|appIndex|executorID|totalCores|maxMem     |maxOnHeapMem|maxOffHeapMem|exec_cpu|exec_mem|exec_gpu|exec_offheap|task_cpu|task_gpu|
-+--------+----------+----------+-----------+------------+-------------+--------+--------+--------+------------+--------+--------+
-|1       |0         |4         |13984648396|13984648396 |0            |null    |null    |null    |null        |null    |null    |
-|1       |1         |4         |13984648396|13984648396 |0            |null    |null    |null    |null        |null    |null    |
++--------+-----------------+------------+----------------+-----------+------------+-------------+-------------+--------------+------------------+---------------+-------+-------+
+|appIndex|resourceProfileId|numExecutors|coresPerExecutor|maxMem     |maxOnHeapMem|maxOffHeapMem|executorCores|executorMemory|numGpusPerExecutor|executorOffHeap|taskCpu|taskGpu|
```
Comment (tgravescs) [inline tools/README.md:260]: What is difference in executor cores and cores per executor?

File: tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/CompareApplications.scala (hunk 3 above)
Comment (tgravescs) [reply tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/CompareApplications.scala:39]: lets do this separately

File: tools/README.md (hunk 9 above)
Comment (nartal1) [reply tools/README.md]: These are same. Removed duplicate  field. PTAL.

**Comment by nartal1** [comment]
build

**Review by tgravescs — APPROVED** [review]

**Pull request merged**